### PR TITLE
Overhaul checkbox element

### PR DESCRIPTION
**Pull request by eliotwrobson**
Remake of #10524 with changes that were not included in #13192. 

Changes:
- Adds grading text generation to mustache template to reduce total LoC
- Changes display to use `display` attribute set to `inline` or `block`.
- Changes order selection to use `order` attribute set to `fixed` or `random`.
- Simplifies partial credit selection logic

Testing: Manually inspected and checked that element renders as expected (including displaying help text that has been migrated into the mustache template), and spot checked that grading worked roughly as expected (though this was not a detailed look, as there are unit tests for specific situations).

Test for reviewer: Manually make sure that old/new attributes behave as expected and that help text behaves as expected. Try a few submissions to make sure the grading algorithm works as expected. Double check that unit tests cover enough backwards compatibility cases.

### CODE REVIEW

File: apps/prairielearn/elements/pl-checkbox/pl-checkbox.py [hunk 1]
```diff
@@ -10,7 +11,7 @@
 
 
 class PartialCreditType(Enum):
-    ALL_OR_NOTHING = "none"
+    OFF = "none"
     NET_CORRECT = "PC"
```
Comment (reteps) [inline apps/prairielearn/elements/pl-checkbox/pl-checkbox.py]: these enum shortcuts need to be renamed

File: apps/prairielearn/elements/pl-checkbox/pl-checkbox.py [hunk 2]
```diff
@@ -438,11 +500,13 @@ def render(element_html: str, data: pl.QuestionData) -> str:
...
     show_answer_feedback = True
-    if (partial_credit and editable) or hide_score_badge:
+    if (
+        partial_credit_mode is not PartialCreditType.OFF and editable
+    ) or hide_score_badge:
         show_answer_feedback = False
 
     display_answers = data["params"].get(name, [])
-    inline = pl.get_boolean_attrib(element, "inline", INLINE_DEFAULT)
+    inline = get_display_type(element) is DisplayType.INLINE
```
Comment (reteps) [inline apps/prairielearn/elements/pl-checkbox/pl-checkbox.py:484]: `is` for enums?

File: apps/prairielearn/elements/pl-checkbox/pl-checkbox.py [hunk 3]
```diff
@@ -712,31 +766,29 @@ def grade(element_html: str, data: pl.QuestionData) -> None:
...
-            number_wrong = len(submitted_set - correct_set) + len(
-                correct_set - submitted_set
-            )
-            score = 1 - 1.0 * number_wrong / number_answers
-        elif partial_credit_method == "COV":
-            n_correct_answers = len(correct_set & submitted_set)
-            base_score = n_correct_answers / len(correct_set)
-            guessing_factor = n_correct_answers / len(submitted_set)
-            score = base_score * guessing_factor
+    # NOTE should probably compute this in a separate function for easier testing
```
Comment (reteps) [inline apps/prairielearn/elements/pl-checkbox/pl-checkbox.py]: Why not do this now? If not, mark this `TODO` not `NOTE`

File: apps/prairielearn/elements/pl-checkbox/test_refactor.py [hunk 4]
```diff
@@ -321,10 +321,40 @@ def build_element(
         == pl_checkbox.PartialCreditType.EACH_ANSWER
     )
 
-    with pytest.raises(ValueError, match="Unknown partial_credit_method"):
+    with pytest.raises(ValueError, match=r"partial-credit-method is deprecated.*"):
         pl_checkbox.get_partial_credit_mode(build_element(True, "INVALID"))
 
 
+def test_partial_credit_new() -> None:
```
Comment (reteps) [inline apps/prairielearn/elements/pl-checkbox/test_refactor.py:364]: There were test files in the other PR. Can you confirm that you don't want any of those tests?

File: apps/prairielearn/elements/pl-checkbox/test_refactor.py [hunk 5]
```diff
@@ -321,10 +321,40 @@ def build_element(
...
+    assert (
+        pl_checkbox.get_partial_credit_mode(build_element("each-answer"))
+        == pl_checkbox.PartialCreditType.EACH_ANSWER
+    )
+
+    with pytest.raises(ValueError, match=r"Invalid partial-credit-method.*"):
+        pl_checkbox.get_partial_credit_mode(build_element("other"))
+
+
 def test_order_type_conversion() -> None:
```
Comment (reteps) [inline apps/prairielearn/elements/pl-checkbox/test_refactor.py:394]: Some explicit confirmation that there are backwards-compatibility tests for everything would be good.

File: apps/prairielearn/elements/pl-checkbox/test_refactor.py (hunk 4 above)
Comment (eliotwrobson) [reply apps/prairielearn/elements/pl-checkbox/test_refactor.py:364]: Yes, as far as I could tell all the tests from the files in the previous PR were included here.

File: apps/prairielearn/elements/pl-checkbox/pl-checkbox.py (hunk 2 above)
Comment (eliotwrobson) [reply apps/prairielearn/elements/pl-checkbox/pl-checkbox.py:484]: I believe this is the recommended comparison: https://docs.python.org/3/howto/enum.html#comparisons

File: apps/prairielearn/elements/pl-checkbox/pl-checkbox.py (hunk 3 above)
Comment (eliotwrobson) [reply apps/prairielearn/elements/pl-checkbox/pl-checkbox.py]: I think it's more useful to make this change along with something else that also modifies the grading algorithm. Happy to make it into a todo but small enough change it's probably better to bundle with something else that will benefit from this.

File: apps/prairielearn/elements/pl-checkbox/test_refactor.py (hunk 4 above)
Comment (reteps) [reply apps/prairielearn/elements/pl-checkbox/test_refactor.py:364]: That's not true, I see test files in #10524 not in this PR.

File: apps/prairielearn/elements/pl-checkbox/pl-checkbox.py (hunk 3 above)
Comment (reteps) [inline apps/prairielearn/elements/pl-checkbox/pl-checkbox.py]:
```suggestion
    # TODO: compute this in a separate function for easier testing
```

File: apps/prairielearn/elements/pl-checkbox/test_refactor.py (hunk 4 above)
Comment (eliotwrobson) [reply apps/prairielearn/elements/pl-checkbox/test_refactor.py:364]: As far as I can tell, the test files themselves are named differently / have slight differences in logic but the tests themselves are the same (though some of them were renamed for whatever reason).

File: apps/prairielearn/elements/pl-checkbox/pl-checkbox.py (hunk 1 above)
Comment (eliotwrobson) [reply apps/prairielearn/elements/pl-checkbox/pl-checkbox.py]: @reteps what should these be changed to?

File: apps/prairielearn/elements/pl-checkbox/test_refactor.py (hunk 5 above)
Comment (eliotwrobson) [reply apps/prairielearn/elements/pl-checkbox/test_refactor.py:394]: Added backwards compatibility tests for the changed attributes: https://github.com/PrairieLearn/PrairieLearn/pull/13324/commits/f58df8be3073069fe0e6340bef74e16defd46bed

File: apps/prairielearn/elements/pl-checkbox/pl-checkbox.py (hunk 1 above)
Comment (reteps) [reply apps/prairielearn/elements/pl-checkbox/pl-checkbox.py]: I'm thinking `none`, `NET`, `EACH`, `COV`

File: apps/prairielearn/elements/pl-checkbox/pl-checkbox.py [hunk 6]
```diff
@@ -238,6 +291,24 @@ def validate_min_max_options(
     Raises:
         ValueError: If any validation constraints are violated
     """
+    if not (0 <= min_correct <= max_correct <= len_correct):
+        raise ValueError(
```
Comment (reteps) [inline apps/prairielearn/elements/pl-checkbox/pl-checkbox.py:287]: The `INTERNAL ERROR` doesn't mean much. Let's get these messages prepared for instructors to read them.

File: apps/prairielearn/elements/pl-checkbox/pl-checkbox.py [hunk 7]
```diff
@@ -238,6 +291,24 @@ def validate_min_max_options(
...
+            f"INTERNAL ERROR: correct number: ({min_correct}, {max_correct}, {len_correct}, {len_incorrect})"
+        )
+
+    min_incorrect = number_answers - max_correct
+    max_incorrect = number_answers - min_correct
+    if not (0 <= min_incorrect <= max_incorrect <= len_incorrect):
+        raise ValueError(
+            f"INTERNAL ERROR: incorrect number: ({min_correct}, {max_correct}, {len_correct}, {len_incorrect})"
```
Comment (reteps) [inline apps/prairielearn/elements/pl-checkbox/pl-checkbox.py]: Internal error message that can be improved.

File: apps/prairielearn/elements/pl-checkbox/test_refactor.py [hunk 8]
```diff
@@ -321,10 +321,40 @@ def build_element(
...
+        pl_checkbox.get_partial_credit_mode(build_element("coverage"))
+        == pl_checkbox.PartialCreditType.COVERAGE
+    )
+    assert (
+        pl_checkbox.get_partial_credit_mode(build_element("each-answer"))
+        == pl_checkbox.PartialCreditType.EACH_ANSWER
+    )
+
+    with pytest.raises(ValueError, match=r"Invalid partial-credit-method.*"):
+        pl_checkbox.get_partial_credit_mode(build_element("other"))
```
Comment (reteps) [inline apps/prairielearn/elements/pl-checkbox/test_refactor.py:391]: We have some code that uses `set(answers)` or `set(correct_answers)`. Can we add a test for cases in which there are duplicates?

**Review by reteps — COMMENTED** [review]
This is looking good! We will merge this after the code freeze.

File: exampleCourse/questions/element/checkbox/question.html [hunk 9]
```diff
@@ -51,11 +51,11 @@
 
 <pl-card header="Part 3">
     <pl-question-panel>
-        <p>Here, the <code>pl-checkbox</code> element displays options as they were written, e.g. <code>fixed-order="true"</code>, and gives more point weight to this question vs. priors, e.g. <code>weight="3"</code>  </p>
+        <p>Here, the <code>pl-checkbox</code> element displays options as they were written, e.g. <code>order="fixed"</code>, and gives more point weight to this question vs. priors, e.g. <code>weight="3"</code>  </p>
```
Comment (reteps) [inline exampleCourse/questions/element/checkbox/question.html:54]: Make sure you update the documentation as well.

**Comment by reteps** [comment]
Also, please resolve any coderabbit comments.

File: exampleCourse/questions/element/checkbox/question.html (hunk 9 above)
Comment (eliotwrobson) [reply exampleCourse/questions/element/checkbox/question.html:54]: https://github.com/PrairieLearn/PrairieLearn/pull/13324/commits/a75b1be5b401839a98c2b4e421166c5c509b2302#diff-ef20c77a77d5b9b229d9f077842a5215a646f9b93fa89a36a3bc9144972188ce

File: apps/prairielearn/elements/pl-checkbox/pl-checkbox.mustache [hunk 10]
```diff
@@ -48,7 +48,31 @@
...
+where <code>t</code> is the number of true options that you select, <code>c</code> is the total number of true options,
+and <code>n</code> is the total number of options you select."
+{{/coverage}}
+
+{{#off}}
+You must select {{{insert_text}}} You will receive a score of 100%
+if you select all options that are true and no options that are false.
+Otherwise, you will receive a score of 0%.
+{{/off}}
 {{/format}}
```
[the remark this answers is not in the document]
Comment (eliotwrobson) [reply apps/prairielearn/elements/pl-checkbox/pl-checkbox.mustache:76]: https://github.com/PrairieLearn/PrairieLearn/pull/13324/commits/7d24fbf7147588532f843e4c0417b79e89152a9a#diff-fbfb9633bcc7636b863669f336e43e4c9824cbdcca99ea2a018c02320056aee2

File: apps/prairielearn/elements/pl-checkbox/pl-checkbox.py (hunk 6 above)
Comment (eliotwrobson) [reply apps/prairielearn/elements/pl-checkbox/pl-checkbox.py:287]: https://github.com/PrairieLearn/PrairieLearn/pull/13324/commits/7d24fbf7147588532f843e4c0417b79e89152a9a#diff-fef85c9cd39a682c06e968b72bfa0bde5996126f82c30d25ba01ed91be42b667

File: apps/prairielearn/elements/pl-checkbox/pl-checkbox.py (hunk 1 above)
Comment (eliotwrobson) [reply apps/prairielearn/elements/pl-checkbox/pl-checkbox.py]: https://github.com/PrairieLearn/PrairieLearn/pull/13324/commits/7d24fbf7147588532f843e4c0417b79e89152a9a#diff-fef85c9cd39a682c06e968b72bfa0bde5996126f82c30d25ba01ed91be42b667

File: apps/prairielearn/elements/pl-checkbox/pl-checkbox.py (hunk 7 above)
Comment (eliotwrobson) [reply apps/prairielearn/elements/pl-checkbox/pl-checkbox.py]: https://github.com/PrairieLearn/PrairieLearn/pull/13324/commits/7d24fbf7147588532f843e4c0417b79e89152a9a#diff-fef85c9cd39a682c06e968b72bfa0bde5996126f82c30d25ba01ed91be42b667

File: apps/prairielearn/elements/pl-checkbox/test_refactor.py (hunk 8 above)
Comment (eliotwrobson) [reply apps/prairielearn/elements/pl-checkbox/test_refactor.py:391]: https://github.com/PrairieLearn/PrairieLearn/pull/13324/commits/a75b1be5b401839a98c2b4e421166c5c509b2302#diff-33e9d5e47fe55ffeba1487be20639c02e35809c9d200e84c141ad5f0b939d82f

File: apps/prairielearn/elements/pl-checkbox/pl-checkbox.py [hunk 11]
```diff
@@ -238,6 +293,24 @@ def validate_min_max_options(
...
+            f"Invalid configuration: The number of incorrect answers needed ({min_incorrect} to {max_incorrect}) exceeds the number of incorrect answers available ({len_incorrect})"
+        )
+
+    if min_select < min_select_default:
+        raise ValueError(
+            f"The attribute min-select is {min_select}, but must be at least {min_select_default}"
+        )
+
+        # Check that min_select, max_select, number_answers, min_correct, and max_correct all have sensible values relative to each other.
     if min_select > max_select:
```
[the remark this answers is not in the document]
Comment (eliotwrobson) [reply apps/prairielearn/elements/pl-checkbox/pl-checkbox.py:304]: This is fine, we only need to give one validation error.

[the remark this answers is not in the document]
Comment (eliotwrobson) [reply apps/prairielearn/elements/pl-checkbox/pl-checkbox.py:304]: This is fine, we only need to give one validation error.

File: apps/prairielearn/elements/pl-checkbox/pl-checkbox.py [hunk 12]
```diff
@@ -712,31 +768,29 @@ def grade(element_html: str, data: pl.QuestionData) -> None:
...
+    elif partial_credit_mode is PartialCreditType.EACH_ANSWER:
+        number_wrong = len(submitted_set - correct_set) + len(
+            correct_set - submitted_set
+        )
+        score = 1 - 1.0 * number_wrong / number_answers
+    elif partial_credit_mode is PartialCreditType.COVERAGE:
+        n_correct_answers = len(correct_set & submitted_set)
+        base_score = n_correct_answers / len(correct_set)
+        guessing_factor = n_correct_answers / len(submitted_set)
+        score = base_score * guessing_factor
```
[the remark this answers is not in the document]
Comment (eliotwrobson) [reply apps/prairielearn/elements/pl-checkbox/pl-checkbox.py:764]: Validated by parse so not a big deal

**Comment by eliotwrobson** [comment]
@reteps now that the deploy freeze is over, can we take a look at getting this merged soon?

File: apps/prairielearn/elements/pl-checkbox/pl-checkbox.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (reteps) [reply apps/prairielearn/elements/pl-checkbox/pl-checkbox.py:304]: The comment is indented, not the code.

File: docs/elements/pl-checkbox.md [hunk 13]
```diff
@@ -50,15 +49,28 @@ a `pl-answer` that has attributes:
 
 ## Partial credit grading
 
-Three grading methods are available when using `partial-credit="true"`:
+Four grading methods are available using the `partial-credit` attribute:
 
-- `"COV"` (Coverage): in this method, the final score is calculated by multiplying the **base score** (the proportion of correct answers that are chosen) with
-  the **guessing factor** (the proportion of chosen answers that are correct). Specifically, if `t` is the number of correct answers chosen, `c` is the total number
-  of correct answers, and `n` is the total number of answers chosen, then the final score is `(t / c) * (t / n)`. This grading scheme rewards submissions that include (i.e. "cover") all true options.
+- `"off"` (All-or-nothing, default): Students receive full credit only if they select all correct answers and no incorrect answers. Otherwise, they receive zero credit.
```
Comment (reteps) [inline docs/elements/pl-checkbox.md:54]: I changed `"off"` to be the default for the partial-credit attribute.

File: docs/elements/pl-checkbox.md [hunk 14]
```diff
@@ -50,15 +49,28 @@ a `pl-answer` that has attributes:
...
+- `"coverage"` (Coverage): The final score is calculated by multiplying the **base score** (the proportion of correct answers that are chosen) with the **guessing factor** (the proportion of chosen answers that are correct). Specifically, if `t` is the number of correct answers chosen, `c` is the total number of correct answers, and `n` is the total number of answers chosen, then the final score is `(t / c) * (t / n)`. This grading scheme rewards submissions that include (i.e. "cover") all true options.
+
+- `"each-answer"` (Every Decision Counts): The checkbox answers are considered as a list of true/false answers. If `n` is the total number of answers, each answer is assigned `1/n` points. The total score is the summation of the points for every correct answer selected and every incorrect answer left unselected.
+
+### Migrating from deprecated attributes
+
+The following deprecated attributes are still supported for backward compatibility:
+
+| Old syntax                                          | New syntax                     |
```
Comment (reteps) [inline docs/elements/pl-checkbox.md:66]: Table mapping between the two for users -- note that in the old system if a user didn't specify any partial credit attributes, it would default to `partial-credit="false"`, so the new default should be `partial-credit="off"`.

This was a bug I fixed -- this would have been bad if it slipped through.

**Comment by reteps** [comment]
### Code review

Found 1 issue:

1. Documentation error - Line 92 in pl-checkbox.md incorrectly states that `max-select` falls back to `min-correct`, but the implementation uses `max-correct`.

https://github.com/PrairieLearn/PrairieLearn/blob/d55ff73fec179e3694a2522a9ae516974b3901a3/docs/elements/pl-checkbox.md#L91-L93

The documentation says:
> If `max-select` is not specified, but `min-correct` is specified along with `detailed-help-text="true"`, then we use the specified value of `min-correct`.

But the code at lines 905-908 of pl-checkbox.py correctly uses `max-correct`:
```python
elif pl.has_attrib(element, "max-correct") and detailed_help_text:
    max_options_to_select = pl.get_integer_attrib(element, "max-correct")
```

🤖 Generated with [Claude Code](https://claude.ai/code)

<sub>- If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

**Review by reteps — APPROVED** [review]
I did some manual testing with `element/checkbox`. I also think the `test_refactor` tests are pretty comprehensive. Let's do it!
